Query loop / Post template: Enable post format filter

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Enables filtering the query loop result by post format.
Partial for https://github.com/WordPress/gutenberg/issues/53049

Additional props @justintadlock @SergeyBiryukov 

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
This feature is often requested by users, see the linked issue.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Description updated September 5 2024.

This PR attempts to enable filtering the post query by post format, but handles the post format **separately from
other taxonomies**, as requested by reviewers.

### REST API changes:
A new class `Gutenberg_REST_Posts_Controller_6_7` that extends `WP_REST_Posts_Controller` is added to the 6.7 compatibility folder.
The intention of this new temporary class is to add 'format' to the Posts controller, so that it can be used to filter the query result per post format.

Updates are made to these class methods: `get_item`, `get_collection_params`

### Query loop block:
- Adds a new parameter, `format`, to the `query` block attribute in the query loop block.
- Adds a new `FormatControls` that uses a list of supported post formats inside a `FormTokenField`.
- `getThemeSupports` is used to get the supported post formats from the store.

### Post Template block:
- The new `format` parameter is passed from the query loop block to the post template block as part of the `query`.
- The post format is added to the `query` using the new `format` key (as opposed to using the post format taxonomy).

### Filtering query_loop_block_query_vars
Filter the query vars (tax_query) for the Query Loop block to support post formats on the front of the site.

## Todo & follow ups
_To be handled as a follow up in a separate PR:_
Ensure that the query loop displays the correct posts when the block is set to **inherit** the query, when the block is placed in a post format archive template.


## Testing Instructions
Enable post formats in a theme by adding the theme support.  Example:
```
function twentytwentyfour_setup(){
	add_theme_support( 'post-formats', array( 'standard', 'aside', 'gallery', 'audio', 'video', 'link', 'image', 'chat', 'status', 'quote' ) );
}

add_action( 'after_setup_theme', 'twentytwentyfour_setup' );
```
Add some test content, both with and without post formats assigned.

Test the query loop both in a template and in a post/page.

Add or select an existing query loop block.
Enable the **custom** query type in the query loop settings sidebar.

Find and open the Filters panel.
Select the three dot menu on the filters panel and toggle all available filters.
Confirm that one of the filters is an input field for "Formats.".
Confirm that all post formats that you registered with the theme supports can be entered into and selected in the input field.

Test that there are no regressions with the query loop features and its filters:
- When the format filter is not enabled
- When the format filter is enabled
- When a single format is selected
-  When multiple formats are selected
- Especially test that the 'standard' format displays posts that has no post format assigned, when it is chosen as the only format, and when combined with another format.
- Confirm that you can reset the format filter with the reset options in the Filter panel.

- Test with classic themes and block themes
- Test that there are no regressions if post formats are not enabled on the theme (All features should work as before this patch).

### Testing custom post types
Follow the steps above to register support for post formats in the theme.

Register a custom post type with post format support, include:
```
'show_in_rest' => true 
'supports' => array( 'title', 'editor', 'thumbnail', 'post-formats'),
```

```
function twentytwentyfour_custom_post_type() {
	register_post_type( 'test',
		array(
			'labels' => array(
				'name' => __( 'Tests' ),
				'singular_name' => __( 'Test' ),
			),
			'public' => true,
			'has_archive' => true,
			'show_in_rest' => true,
			'supports' => array( 'title', 'editor', 'thumbnail', 'post-formats'),
			'taxonomies' => array( 'category', 'post_tag' ),
			'rewrite' => array('slug' => 'tests'),
		)
	);
}

add_action( 'init', 'twentytwentyfour_custom_post_type' );
```

In the admin, create a few items of this post type, with and without a post format assigned.
Go to Appearance > Editor and select a template with a query -or create one.

1. In the query loop setting, change query type to custom.
2. Under post type, select your post type.
3. Open the Filters panel. Confirm that you can toggle the display of the Formats control.
4. When you select / focus on / place the cursor on the input for the formats control there should be a dropdown with the same formats that you registered with the theme support.
5. Confirm that entering a random value that is not a registered format does nothing.
6. Select a format. Confirm if the query loop displays the correct items in the editor and front.
7. Select multiple formats. Confirm if the query loop displays the correct items in the editor and front.

### Switching post types (After registering your custom post type)
In the settings sidebar of the query loop, set the post type to posts.
Open the Filters panel, and add a post format.
Switch the post type to the custom post type. Confirm that the filter is unchanged.
Switch the post type to page, which does not support post formats: Confirm that the filter is not showing in the panel.
Switch the post type to posts: confirm that the filter is reset (it was reset when you switched to the page).

### Testing the query pagination
After creating a number of posts with formats, set the "items per page" setting to something low so that the query pagination block will show on the front.
Save and view the front of the site. Use the pagination and confirm that page 2 etc displays the correct posts.

## Screenshots or screencast <!-- if applicable -->
I apologise, I have compressed the video but it is still almost 4 MB.
In this video I am testing different post formats in a template in the Site Editor.
I am testing formats in combination with other taxonomies, entering a non-existing format name in the control, etc.
I am switching between Post, page, and a custom post type with post format support, to illustrate that the filter is reset when switching to pages, but maintained when switching between posts and custom post types.

(Aside: I am not using a mouse and I do see that I often miss click when trying to remove a format from the control)

https://github.com/user-attachments/assets/9dc8597b-9956-4f13-aa91-bf0030484fa1

carolinan committed:
Enable post format filter in the query loop

carolinan committed:
try to fix CS issues

carolinan committed:
how about this cs fix then?

carolinan commented:
> The PR adds a temporary filter that sets show_in_rest to true. This is added to a new file in the 6.7 compatibility folder.

This also enabled Format Archives as an option in the template creation flow ( Site Editor > Templates > Add New Template.)

youknowriad commented on lib/compat/wordpress-6.7/taxonomy.php:
```diff
@@ -0,0 +1,22 @@
+add_filter(
+	'register_post_format_taxonomy_args',
+	function ( $args ) {
+		$args['show_in_rest'] = true;
```
I wonder what side effects this could have. Looks ok to me but some folks familiar with post formats and/or taxonomies might want to weight in here.

youknowriad reviewed (commented):
What's a bit weird to me in the experience proposed here is the following:

 - To pick a post format for a post, we have a dedicated UI component that is separate from the taxonomies UI.
 - To filter a query loop using a post format, we don't have any custom UI, we just rely on the taxonomies UI.

What I'm wondering is:

Why the taxonomy show up in the query loop filter but doesn't show up in the document sidebar when creating a new post.

youknowriad commented:
Ok I think I know the answer to my question: The post format taxonomy has `show_ui=false` but this flag is only used for the document sidebar panels.

I think I see two action items here:

 - Add the `show_ui` check to the taxonomies shown in the query loop filters. In other words, we should probably hide these taxonomies in this UI
 - Build a dedicated UI to filter per post format in the the query loop filter for post formats (like we have a custom UI for the document sidebar)

In other words, we can't treat this property as a taxonomy in one case but not the other, we should be consistent and I feel like historically, we don't treat this as a taxonomy, we only use the taxonomy as a storage mechanism.

carolinan commented:
The benefit of using the taxonomy filter like now is that it only requires a small code change.

I think the control should still be under filters, and perhaps use a selectControl? It is not possible to register custom post formats, so selecting from a list of supported formats, rather than typing in a value makes sense.

What would need to be fixed for a first iteration though?

carolinan committed:
Create a new option for post formats

carolinan commented:
Ok, I have pushed changes on the editor side, moving the post formats out of the taxonomies.
Since the formats are no longer part of the `tax_query`, it will require changes to `build_query_vars_from_query_block` to work on the front.

I am still not sure that moving it out of the taxonomy is worth this amount of code.

carolinan commented:
Another question: Is it expected that only post formats that are used can be selected in the query loop option?
I mean that I have registered theme support for multiple formats, but only assigned one post to the video format, and the video is the only format I can select in the query.

youknowriad commented on packages/block-library/src/query/edit/inspector-controls/post-format-controls.js:
```diff
@@ -0,0 +1,51 @@
+		( select ) => {
+			const { getEntityRecords } = select( coreStore );
+			return {
+				postFormats: getEntityRecords( 'taxonomy', 'post_format', {
```
So looking at the existing PostFormat control from the inspector control (post-format-controls.js). It seems like it's not using the REST API for this (or at least not this endpoint)

It's using `const supportedFormats = select( coreStore ).getThemeSupports()?.formats;`
and using that array to filter a static list of post formats defined on that file.

So my thinking is, maybe we should be consistent and do the same here as well. (We could potentially extract the list of formats to the `constants.js` file of the package and use it in both places.

I'm guessing this answers the question that you asked about the available post formats to pick from.

I'm guessing this also means that we don't really need to make the post format taxonomy's show_in_rest true anymore.

carolinan replied:
I'm going to try it, but need to figure out how to display the correct formats per post type, not only list all the formats available in the theme.
If the theme supports all, but a **custom** post type only support one, only one should show.

carolinan replied:
When a post is assigned a format WP saves a term taxonomy id that needs to be used in the query to get the correct content for the preview in the post template.
getThemeSupports only returns the array of names, not id's.
So if getThemeSupports is used, at some point there still needs to be a second database search that compares the name `aside` with the name or slug in wp_terms `post-format-aside` and gets the matching term id.

carolinan replied:
> display the correct formats per post type,

It seems the block editor does not support the classic PHP filters, so I'm removing this from the todo list (for now).

youknowriad replied:
> getThemeSupports only returns the array of names, not id's

I'm I understanding correctly than when assigning a post format to a "post" object, we use the "names" and not the "ids" ? If that's the case, it seems logical that the REST API to filter posts should support a dedicated postFormat argument that accept names and not ids.

carolinan replied:
I am trying to understand that too:

https://developer.wordpress.org/rest-api/reference/posts/#schema
https://developer.wordpress.org/rest-api/reference/posts/#schema-format
Expects a string with a name.

https://developer.wordpress.org/reference/functions/wp_set_post_terms/
Expects a string or array with either an ID or name.

"An array of terms to set for the post, or a string of terms separated by commas. Hierarchical taxonomies must always pass IDs rather than names so that children with the same names but different parents aren’t confused."

carolinan replied:
I don't know how else to fix the rest API error or the PHP error described in the description of the PR (Under "How?")

carolinan replied:
If I only
1) Check the length of the postFormat attribute
2) Convert that to a comma separated string
3) Add that to the query

I get:

1) The spinner in the preview
2):

`{"code":"rest_invalid_param","message":"Invalid parameter(s): post_format","data":{"status":400,"params":{"post_format":"post_format is not a valid Term ID List. Reason: post_format[0] is not of type integer."},"details":{"post_format":{"code":"rest_no_matching_schema","message":"post_format is not a valid Term ID List. Reason: post_format[0] is not of type integer.","data":{"position":0}}}}}
`

youknowriad replied:
Thought issue indeed., I think for the whole PR we need to answer this question: Should we use post_format as a taxonomy or not? For me the answer is probably **no** given how the "post" endpoint returns the `format` property for each post and allows editing the `format` as well by passing a string.  cc @TimothyBJacobs for opinion as well. 

This means to me that ideally the `/posts` endpoint would allow passing `format` as a string in the query rather than a list of ids. 

Now, looking at this PR, it seems that we already support `post_format` in the query? and it needs to be integers. Is that correct? When was this implemented (I don't see it on the REST API handbook)

youknowriad replied:
Is the `post_format` in the query supported because you added `show_in_rest` to the taxonomy?

carolinan replied:
I don't fully understand how it works, but yes. I am not even entirely sure which end point is being used and where I can see the code of it!

I tried disabling show_in_rest.
I also tried using 'format' instead of 'post_format'.
It did not work, with numeric id or string.

youknowriad replied:
Ok, here's my current understanding:

 - `show_in_rest` enables access to the taxonomy in the taxonomy endpoint (using `getEntityRecords` with the `post_format` taxonomy).
 - `show_in_rest` also enables filtering the `/posts` endpoint using the taxonomy filter (using `getEntityRecords` with `post` postType and passing `post_format` in the query)

The problem for me is that this means we'll be treating the post format as a regular taxonomy but it's not the case 100% since the post object returns a `format` property in the post object.

I think the idea for me is:

 - Keep `show_in_rest` disabled for post formats.
 - Add support for a new `format` query key to the `posts` endpoint to match how the "posts" already represent the format. This likely means we need to override the Post endpoint controller here to add this and use this property in the block to filter per format. (should accept a string or a list of strings, just like the "format" property on the post object)

I would love some confirmation from folks familiar with these endpoints though. I don't want to suggest something that end up being the wrong thing :)

ntsekouras replied:
I read a bit about the post formats and this last suggestion from Riad not to treat formats as a taxonomy seems a good one.

carolinan replied:
I don't understand how to solve this.

What is "a new format query key"?  Do you mean it needs to be in get_collection_params?
https://github.com/WordPress/wordpress-develop/blob/6.6/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2814

>Override the Post Endpoint Controller.

Does this mean something else than creating a new class, extending WP_REST_Posts_Controller?

ntsekouras replied:
>What is "a new format query key"? Do you mean it needs to be in get_collection_params?

Yes, among some other changes that are going to be needed with an example to update the `get_items` function to check for the new query arg and take it into account to return the proper results.

carolinan replied:
I have tried, but it is beyond my abilities.

carolinan replied:
I have got something working, but I am expecting it to not pass code review 😅

carolinan replied:
"got something working"  -except now the query has no results when no formats are used 👎

nickbohle commented:
> Another question: Is it expected that only post formats that are used can be selected in the query loop option? I mean that I have registered theme support for multiple formats, but only assigned one post to the video format, and the video is the only format I can select in the query.

I would favour the idea of being able to select one of all registered post formats. If there is no video format post, the result of the query loop would be no posts found. But you would be able to design your site / video format query loop from scratch, even if no such posts exists yet.

carolinan committed:
Update the control to use the post format theme support

carolinan committed:
Post template preview: Get the ID of the post format to use in the query.

carolinan commented:
I have pushed another update, but I really do not like these changes to post-template/edit.js.

youknowriad commented on packages/block-library/src/post-template/edit.js:
```diff
@@ -138,7 +138,10 @@ export default function PostTemplateEdit( {
 						if ( taxonomy?.rest_base ) {
-							accumulator[ taxonomy?.rest_base ] = terms;
+							// Skip post formats, as they are handled separately.
+							if ( ! taxonomy.rest_base === 'post_format' ) {
```
I don't like this change as well, I think we should skip taxonomies with `show_ui` false like we do for the regular taxonomies control.

carolinan replied:
I don't understand. This is the post preview, not the control.

youknowriad commented on packages/block-library/src/post-template/edit.js:
```diff
@@ -163,6 +166,54 @@ export default function PostTemplateEdit( {
+					return { posts: [] };
+				}
+
+				query.post_format = formatIds;
```
I'm not sure I understand why we need all the logic here, can't we assign this automatically? Why do we need to do a request to get the right argument?

carolinan committed:
Filter the query loop vars to support post formats on the front.

carolinan committed:
Remove the post type from PostFormatControls

carolinan committed:
PHPCS fixes

nickbohle commented:
> Maybe we can ask for other opinions.

Okay, my two cents. Since their introduction I am using post formats as a third taxonomy. With their own slug /type: besides /category and /tag.

It all started with Twenty-Fourteen for me, when I decided to only show standard posts. I found a solution by WordPress hero's like Niacin.

This was the past, we will win.

TimothyBJacobs commented:
> For me the answer is probably no given how the "post" endpoint returns the format property for each post and allows editing the format as well by passing a string. cc @TimothyBJacobs for opinion as well.

> This means to me that ideally the /posts endpoint would allow passing format as a string in the query rather than a list of ids.

Agreed on both counts here. I think this is also evident from a PHP API perspective where we have helper functions to interact with post formats like `set_post_format`.

nickbohle commented:
> Maybe we can ask for other opinions.

I’ve been using Post Formats since its introduction and with
- the slug /type/
- Twenty-Fourteen and the 'Ephemera' widget
- a code snippet / plugin to generate the ‚standard‘ post format

as a taxonomy. Why? I just wanted to separate blog posts from great food pictures or quotes I’ve just read.

And yes, post formats are also the tool to translate WordPress posts into ActivtyPub objects.

So going forward, when I make an aside post for the Fediverse, I want to be able to exclude it from my standard blog posts.

ntsekouras commented on lib/compat/wordpress-6.7/blocks.php:
```diff
@@ -43,3 +43,33 @@ function gutenberg_filter_block_type_metadata_settings_allow_variations_php_file
+			$block->context['query']['postFormat']
+		);
+
+		$query['tax_query'] = array(
```
By doing that here, we always override other taxonomy filters and only handle the post format one.

ntsekouras commented on packages/block-library/src/query/edit/inspector-controls/index.js:
```diff
@@ -52,6 +53,7 @@ export default function QueryInspectorControls( props ) {
 		inherit,
 		taxQuery,
 		parents,
+		postFormat,
```
We also need to update `onPostTypeChange` below, to update the applied formats, if any..

carolinan replied:
Fixed to the best of my ability, it seems to work.

ntsekouras commented on packages/block-library/src/query/edit/inspector-controls/index.js:
```diff
@@ -128,11 +130,18 @@ export default function QueryInspectorControls( props ) {
+	// TODO: this condition does not work for custom post types,
+	// useTaxonomies util does not return the post format taxonomy
+	// for custom post types, only posts?
+	const showPostFormatControl =
```
Have you tested by adding the support in the custom post type? By reading this it seems it's not supported by default..

carolinan replied:
Yes, I registered the support. The post format shows in the format option in the block editor.

carolinan replied:
I have replaced this with a check for the theme support.

carolinan committed:
Merge branch 'trunk' into try/post-format-query

carolinan committed:
Update the condition for when then post format control should show

- Update the condition for when then post format control should show, using the core store to get the theme supports.

-Remove the post format check from the taxQuery build.

carolinan committed:
WIP: Try to add format to the REST API posts controller through a custom Class.

carolinan committed:
Merge branch 'trunk' into try/post-format-query

carolinan committed:
Update since

Add the changes to the doc block of each method.

carolinan committed:
Regenerate the fixture for the query loop, including postFormat.

carolinan committed:
Revert removing the post formats from TaxonomyControls

carolinan committed:
Update class-gutenberg-rest-posts-controller-6-7.php

carolinan commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,1139 @@
+			);
+		}
+
+		/*
+		* I have temporarilly removed this because I believe it is incorrect,
+		* if I don't comment this out, then there are no results if
+		* the query loop block is not set to filter by format.
+		$query_params['format'] = array(
+			'default'     => 'standard',
+			'description' => __( 'Limit result set to posts assigned one or more formats.' ),
+			'type'        => 'array',
+			'items'       => array(
+				'anyOf' => array(
+					array(
+						'type' => 'string',
+						'enum' => array_values( get_post_format_slugs() ),
+					),
+					array(
+						'type'  => 'array',
+						'items' => array(
+							'type' => 'string',
+							'enum' => array_values( get_post_format_slugs() ),
+						),
+					),
+				),
+			),
+		);
+		*/
+
```
This code does not seem to be needed, rather, it caused an issue where if I did not select any filters in the query loop block, there were no results. The expected result was for the latests posts to be displayed.

When removed/ commented out, the _filter_ still seems to work in the editor.

I don't know the REST API and the posts control well enough to know what updates are needed exactly, so this was part of my experimentation.

carolinan commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,1139 @@
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return array Updated query arguments.
+	 */
+	private function prepare_tax_query( array $args, WP_REST_Request $request ) {
```
No changes were made here, the function is only included because it is private in the original class...

carolinan commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,1139 @@
+	 * @param array $query_params Collection schema.
+	 * @return array Updated schema.
+	 */
+	private function prepare_taxonomy_limit_schema( array $query_params ) {
```
No changes were made here, the function is only included because it is private in the original class...

carolinan commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,1139 @@
+			),
+		);
+
+		$schema['properties']['format'] = array(
```
The change starts here.
I am trying to allow both a string or an array, but only the valid post format slugs.
It might be fine to just allow the string?

carolinan replied:
Update: There is definitely something wrong, because if I don't comment this out, I am not able to assign formats to posts in the block editor.

carolinan commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,1139 @@
+
+		$args = $this->prepare_tax_query( $args, $request );
+
+		if ( isset( $request['format'] ) ) {
```
This is the only change to get_items.

carolinan commented on packages/block-library/src/query/block.json:
```diff
@@ -25,7 +25,8 @@
 				"taxQuery": null,
-				"parents": []
+				"parents": [],
+				"postFormat": []
```
Perhaps this should be a string.

carolinan committed:
Combine the post format with the tax_query array instead of replacing it.

carolinan committed:
WIP: comment out breaking code to allow testing.

carolinan committed:
Merge branch 'trunk' into try/post-format-query

carolinan committed:
edit.js: Don't convert the post format to a string

Post template edit.js: Don't convert the post format to a string. This conversion prevented selecting multiple post formats in the query filter.

carolinan committed:
Update edit.js

Mamaduka commented on lib/compat/wordpress-6.7/rest-api.php:
```diff
@@ -10,6 +10,18 @@
+ * to add support for post formats.
+ * - Hardcoded the 'post' post type for now.
+ */
+function gutenberg_post_format_rest_posts_controller() {
+	$posts_controller = new Gutenberg_REST_Posts_Controller_6_7( 'post' );
+	$posts_controller->register_routes();
+}
+
+add_action( 'rest_api_init', 'gutenberg_post_format_rest_posts_controller' );
```
I think the preferred way of doing this is using the `register_post_type_args` filter.

```php
function gutenberg_post_format_rest_posts_controller( $args, $post_type ) {
	if ( 'post' === $post_type ) {
		$args['rest_controller_class'] = 'Gutenberg_REST_Posts_Controller_6_7';
	}

	return $args;
}
add_filter( 'register_post_type_args', 'gutenberg_post_format_rest_posts_controller', 10, 2 );
```

carolinan replied:
I don't know, so yes, I am open to changing it.
The question remains how to support post formats for custom post types.

Mamaduka replied:
Good questions! We could change the check to something like this:

```php
function gutenberg_post_format_rest_posts_controller( $args, $post_type ) {
	if ( post_type_supports( $post_type, 'post-formats' ) ) {
		$args['rest_controller_class'] = 'Gutenberg_REST_Posts_Controller_6_7';
	}

	return $args;
}
add_filter( 'register_post_type_args', 'gutenberg_post_format_rest_posts_controller', 10, 2 );
```

carolinan replied:
oh yes that makes sense, I will try it, thank you.

carolinan committed:
Simplify the format array in get_item_schema

Change the type to a string that matches the format slugs.
Each post can only have one format.

carolinan commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,1127 @@
+			),
+		);
+
+		$schema['properties']['format'] = array(
+			'description' => __( 'The format for the post.' ),
+			'type'        => 'string',
+			'context'     => array( 'view', 'edit' ),
+			'enum'        => array_values( get_post_format_slugs() ),
+		);
```
This is the only change to get_item_schema.

Mamaduka replied:
The `format` property is only included in the schema when the post type supports it above (see `case 'post-formats`). So why do we need unconditional addition here?

carolinan replied:
I don't have a good enough understanding of this. I thought the` $post_type_attributes` and the `format` needed to be separated.

If this needs to be removed, then I don't think `get_item_schema` needs to be in the class extension at all.
Or are you asking me to make it conditional? (if the post type supports post formats) 
Or are you asking me to move it? Add format to post type attributes?

Mamaduka replied:
If you enable `post-format` for `page` or any custom post type on the trunk, can you set/update it via the post editor?

* If yes, then this code isn't needed.
* If no, this means that `post-formats` are only served for posts, and allowing it for other post types is a matter of different discussion.

carolinan replied:
I can set and update it. :)

carolinan committed:
Remove the changes to get_collection_params()

I could not get this change to work. Maybe it should not even be added, so I am removing it.

carolinan committed:
Allow filtering for the standard post format

carolinan committed:
Filter register_post_type_args instead of using rest_api_init

Mamaduka commented:
I'm sorry if this has already been discussed, but maybe we should standardize post-format parameter names between the REST API and the editor.

The tags' taxonomy name also has a `post` prefix but is exposed via REST API as `tags` when listing a post or retrieving a single post. The post format is also provided via a `format` key.

How about using `format` and `formats` parameters?

carolinan commented:
There are some issues that I will need to leave for another day (likely Monday)
- Displaying the "standard" format on the front.
- Styling the select control.
- Updating onPostTypeChange.
- Standardize post-format parameter names

carolinan committed:
Merge branch 'trunk' into try/post-format-query

carolinan committed:
Rename "postFormat" parameter to "format"

Rename "postFormat" parameter to "format".

Simplify the name of the control from PostFormatControls to FormatControls.

Simplify the name of the constant that checks if the FormatControls should be used, from showPostFormatControl to showFormatControl.

carolinan commented:
>  Displaying the "standard" format on the front.

I think the limitation is in the blocks.php file,  gutenberg_filter_query_loop_block_query_vars_post_format, since if the `standard ` format is passed to it, it of course can't get that format from the database because it does not exist. 
My intention is to only show posts that has no format in this case.

Mamaduka commented on lib/compat/wordpress-6.7/blocks.php:
```diff
@@ -43,3 +43,33 @@ function gutenberg_filter_block_type_metadata_settings_allow_variations_php_file
+		! empty( $block->context['query']['format'] ) ) {
+		// Add the required "post-format-" prefix to each format.
+		$formats = array_map(
+			function ( $format ) {
```
```suggestion
			static function ( $format ) {
```

I think even closures outside classes should be `static`.

Mamaduka commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,937 @@
+
+		$args = $this->prepare_tax_query( $args, $request );
+
+		if ( isset( $request['format'] ) && ! empty( $request['format'] ) ) {
```
Just `! empty` check should be enough here.

Mamaduka commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,937 @@
+			if ( ! empty( $formats ) ) {
+				// Add the post-format- prefix.
+				$terms = array_map(
+					function ( $format ) {
```
```suggestion
					static function ( $format ) {
```

Micro-optimization: Closures/anon-functions in classes should use `static` if they're not accessing `$this`.

Mamaduka commented on packages/block-library/src/query/edit/inspector-controls/format-controls.js:
```diff
@@ -0,0 +1,68 @@
+	];
+
+	return (
+		<SelectControl
+			__nextHasNoMarginBottom
+			__next40pxDefaultSize
+			multiple
+			label={ __( 'Formats' ) }
+			value={ format }
+			options={ formatOptions }
+			onChange={ ( value ) => {
+				onChange( { format: value } );
+			} }
+		/>
```
Is this a temporary placeholder component? Are we planning to switch to the `FormTokenField` later?

carolinan replied:
I don't think this should use the FormTokenField because the options are always the ones supported by the theme.
The number of options is low, and I would like to know if the format is supported or not, without typing the name of the format into an input?

Mamaduka replied:
That should be possible by displaying suggestions (supported formats) on focus. See https://wordpress.github.io/gutenberg/?path=/story/components-formtokenfield--dropdown-selector.

Mamaduka reviewed (commented):
Thanks for standardizing the naming convention, @carolinan!

Is there a reason to use `format` instead of `formats` when an array for post formats is used for querying a post collection? It's not a blocker, but I've seen plural forms used in similar cases.

We should also update `get_collection_params` and list formats when the post type supports it. The [post-status schema ](https://github.com/WordPress/wordpress-develop/blob/e569967ee45e23186f6aa7f479fcb48a7511436b/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2962-L2971) could be used as a blueprint for the new value.

carolinan committed:
Merge branch 'trunk' into try/post-format-query

carolinan committed:
Add format to get_collection_params()

carolinan committed:
Update lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php

Co-authored-by: George Mamadashvili <georgemamadashvili@gmail.com>

carolinan committed:
Update class-gutenberg-rest-posts-controller-6-7.php

carolinan committed:
Merge branch 'try/post-format-query' of https://github.com/WordPress/gutenberg into try/post-format-query

carolinan committed:
PHPCS: remove white space

carolinan committed:
Remove get_item_schema()

carolinan committed:
Update the logic for the standard post format in  gutenberg_filter_query_loop_block_query_vars_post_format

Mamaduka commented on packages/block-library/src/query/edit/inspector-controls/format-controls.js:
```diff
@@ -0,0 +1,68 @@
+		supportedFormats.includes( item.slug )
+	);
+
+	const formatOptions = [
+		...( formats || [] ).map( ( item ) => ( {
+			value: item.slug,
+			label: item.name,
+		} ) ),
+	];
```
This transformation won't be necessary if we change `slug` to `value` and `name` to `label` in the `POST_FORMATS` definition.

Mamaduka commented on packages/block-library/src/query/edit/inspector-controls/format-controls.js:
```diff
@@ -0,0 +1,68 @@
+	{ slug: 'status', name: __( 'Status' ) },
+	{ slug: 'video', name: __( 'Video' ) },
+].sort( ( a, b ) => {
+	const normalizedA = a.name.toUpperCase();
+	const normalizedB = b.name.toUpperCase();
+
+	if ( normalizedA < normalizedB ) {
+		return -1;
+	}
+	if ( normalizedA > normalizedB ) {
+		return 1;
+	}
+	return 0;
```
I'm guessing this sorting is for localized labels.

carolinan replied:
Yes

carolinan replied:
Its from the editor PostFormat component.

carolinan committed:
Replace SelectControl with FormTokenField

carolinan commented:
I have replaced the SelectControl with the FormTokenField, but I don't think it is the best control for this low number of set items.
The small win with using it, is that it looks similar to the rest of the controls in the panel.

Todo:
- [ ]  Updating onPostTypeChange.
- [ ]  Leaarn if I need to add **sanitization** of the user input for the FormTokenField. I left this out today because of lack of time.
- [ ] Update the PR description, media, and testing instructions.

carolinan commented:
> How about using `format` and `formats` parameters?

> Is there a reason to use format instead of formats when an array for post formats is used for querying a post collection? It's not a blocker, but I've seen plural forms used in similar cases.

I do not understand _when_ you are expecting _which_ of the forms (singular or pluralI) to be used 😞

carolinan committed:
Merge branch 'trunk' into try/post-format-query

carolinan committed:
Set the constant format to an empty array when format is not queried

carolinan committed:
Hide the control and reset the format value if the post type does not support post formats.

carolinan committed:
Rename postFormatSupportMap to postTypeFormatSupportMap

carolinan commented:
Now custom post types with formats are not filtered in the editor, only on the front, and I don't know what changed.

carolinan committed:
Merge branch 'trunk' into try/post-format-query

carolinan commented:
> Now custom post types with formats are not filtered in the editor, only on the front, and I don't know what changed.

I suppose I will start by adding the get_item_schema back and see what happens.

Mamaduka commented:
> Now custom post types with formats are not filtered in the editor, only on the front, and I don't know what changed.

Can you provide reproduction steps for this?

carolinan committed:
Merge branch 'trunk' into try/post-format-query

carolinan commented:
> > Now custom post types with formats are not filtered in the editor, only on the front, and I don't know what changed.
> 
> Can you provide reproduction steps for this?

I will add it to the testing description above.

carolinan commented:
Testing instructions have been updated.

carolinan commented:
So, I added an error_log inside `Gutenberg_REST_Posts_Controller_6_7` that outputs  `$this->post_type` and even if I select the custom post type in the editor, the log shows 'post'.

carolinan commented:
Update: rather it does not log **anything** until I switch the post type in the query loop setting to 'post'

carolinan commented:
> I suppose I will start by adding the get_item_schema back and see what happens.

This made no change. I need help from someone who has a deeper understanding of how this all works together/ is supposed to work 😞

Mamaduka commented:
So that we can cover all the bases here, do the other post types you're testing support post formats?

P.S. I can have a better look later today.

carolinan commented:
I went back and tested old commits between now and `63bad0fd40` and the formats filter did not work for custom post types, the wrong posts were shown in the editor. So it probably never worked, my mind is just playing tricks.

Mamaduka committed:
Fix REST API controller override condition

Mamaduka committed:
Don't set the default post format in the schema

It will be used as the default argument if none is supplied.

Mamaduka committed:
Satisfy PHP linter

Mamaduka commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -485,7 +485,6 @@ public function get_collection_params() {
 
 		if ( post_type_supports( $this->post_type, 'post-formats' ) ) {
 			$query_params['format'] = array(
-				'default'     => 'standard',
```
I think when no formats are set, all formats should be displayed. Removing the default argument here solves that.

Mamaduka committed:
Update the 'FormatsControl' to match suggestion labels to their static values

Mamaduka committed:
Refactor showFormatControl logic

dmsnell commented on lib/compat/wordpress-6.7/blocks.php:
```diff
@@ -43,3 +43,58 @@ function gutenberg_filter_block_type_metadata_settings_allow_variations_php_file
+ */
+function gutenberg_filter_query_loop_block_query_vars_post_format( $query, $block ) {
+	if ( isset( $block->context['query']['format'] ) &&
+		! empty( $block->context['query']['format'] ) ) {
```
`empty()` does not generate a warning if the variable doesn't exist, so I think the call to `isset()` here is superfluous

carolinan replied:
Thank you, I will remove it.

dmsnell commented on lib/compat/wordpress-6.7/blocks.php:
```diff
@@ -43,3 +43,58 @@ function gutenberg_filter_block_type_metadata_settings_allow_variations_php_file
+		if ( count( $tax_query ) > 1 ) {
+			$query['tax_query'][] = $tax_query;
+		}
+	}
```
given that this entire `if` nesting is there to ensure the variable exists, it could be nice to flip the condition and early-abort if it doesn't.

```php
if ( empty( $block['context']… ) {
	return $query;
}

// No more nesting is required in this function.
```

carolinan replied:
Updated, thank you 👍

dmsnell commented on lib/compat/wordpress-6.7/blocks.php:
```diff
@@ -43,3 +43,58 @@ function gutenberg_filter_block_type_metadata_settings_allow_variations_php_file
+			);
+			// Remove the standard format:
+			$formats = array_diff( $formats, array( 'standard' ) );
+		}
```
I find the comments here a bit confusing. It seems like we're saying that a query with standard _must_ list it as excluded, but then assuming also that if it's in here, it _is_ excluded.

Does this handle cases where someone requests _including_ `standard`?

carolinan replied:
Standard includes all items that do not have a post format assigned.
It excludes all items that do have a post format assigned.

Standard can not be searched for in the database, that is why it is removed from the query.

When the user selects both standard and a format, they are combined so that you can query for both
"video" and "no format assigned".

carolinan replied:
How about

```
	// The default post format, 'standard', is not stored in the database.
	// If 'standard' is part of the request, the query needs to exclude all post items that
	// has a format assigned.
	if ( in_array( 'standard', $formats, true ) ) {
		$tax_query[] = array(
			'taxonomy' => 'post_format',
			'field'    => 'slug',
			'terms'    => array(),
			'operator' => 'NOT EXISTS',
		);
		// Remove the standard format, since it cannot be queried.
		$formats = array_diff( $formats, array( 'standard' ) );
	}
```

dmsnell replied:
the comment here is much clearer. thank you.

small nitpick: there's a correspondence typo in "all post items that <del>has</del><ins>have</ins> a format assigned."

dmsnell commented on lib/compat/wordpress-6.7/blocks.php:
```diff
@@ -43,3 +43,58 @@ function gutenberg_filter_block_type_metadata_settings_allow_variations_php_file
+				'operator' => 'NOT EXISTS',
+			);
+			// Remove the standard format:
+			$formats = array_diff( $formats, array( 'standard' ) );
```
it would probably be fine to simply `unset( $formats['standard'] )` here instead of calling the `array_diff()` function, given that we know the key we want to remove.

carolinan replied:
`standard` is a value so unset alone does not work.

```suggestion
			unset( $formats[ array_search( 'standard', $formats, true ) ] );
```

dmsnell commented on lib/compat/wordpress-6.7/blocks.php:
```diff
@@ -43,3 +43,58 @@ function gutenberg_filter_block_type_metadata_settings_allow_variations_php_file
 add_filter( 'block_type_metadata_settings', 'gutenberg_filter_block_type_metadata_settings_allow_variations_php_file', 10, 2 );
+
+/**
+ * Filter the query vars (tax_query) for the Query Loop block to support post formats.
```
I see three separate terms in this function and filter:

 - query vars
 - tax query
 - post formats

I would have expected the name and description to match, but I'm left wondering what I miss when I read `tax_query` - why is it there? why does it matter? did I miss something?

Perhaps there's a way to focus on what this method _does_ specifically from the outside, unless it's destined to be a place that many query vars are filtered. This example is probably wrong because I don't understand what the goal is well enough.

For example…
```php
/**
 * Adds appropriate `post_format` filters to the query loop if necessary for
 * a given block type; e.g. if the block only exists in a template.
 *
 * @param …
 * …
 */
function gutenberg_set_query_vars_for_post_format_from_block_context( $query, $block ) {
	…
}
```

from reading this code, I think there could be more explanation of the purpose of this function and the relationship between the query vars and the block and its context.

carolinan replied:
Since it is filtering, I absolutely think that `filter` needs to be part of the function name.

carolinan replied:
Are you asking why `$query['tax_query']` is in the code or why `tax_query` is in the description?

carolinan replied:
```
Filters the taxonomy query vars (`$query['tax_query']`) on the query loop block
to include or exclude posts that has a post format assigned.
```

dmsnell replied:
I think I understand this better now, and I think some of the confusion stems from the fact that there's just so much in the name.

> Since it is filtering, I absolutely think that filter needs to be part of the function name.

This method itself isn't filtering, is it? It's adding additional specifiers to the loop query based on attributes in the query block.

> Are you asking why $query['tax_query'] is in the code or why tax_query is in the description?

My question was less about `tax_query` and more about harmony between the function's description, name, and code.

---

The new description is clearer to me, someone coming into this code without the context of having worked with it already.

Still, I wonder if this could be even more guiding to an unfamiliar audience, without complicating the public interface with internal details.

```php
/**
 * Adds post format query vars to query loop block's WP_Query when the block's attributes call for them.
 *
 * @see 'query_loop_block_query_vars'
 *
 * …
 */
function gutenberg_add_appropriate_post_format_specifiers_to_query_loop( $existing_wp_query, $query_loop_block ) {
```

carolinan replied:
How about just
`gutenberg_add_format_query_vars_to_query_loop`

I am not sure about using "post_format' since this PR was created because it needed to be separated from the post_format taxonomy query.

dmsnell replied:
that name sounds clear and to the point. 👍

dmsnell commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,699 @@
+	 * @since 4.7.0
+	 * @var string
+	 */
+	protected $post_type;
```
this var already appears to be part of `WP_REST_Posts_Controller` - why is being added in this subclass?

Mamaduka replied:
That was my addition to satisfy the IDE PHP linter. Happy to revert if it's confusing. Ref: 8eab071a5ae68e6ea7fb43bf13834b10282db7b7.

dmsnell replied:
ah, your IDE wasn't aware of the base class? I guess it's not _wrong_ to redeclare it here. I'm not familiar enough to know if there are concerns about doing so; I just assumed there might be if we aren't changing its visibilty.

dmsnell commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,699 @@
+
+		$args = $this->prepare_tax_query( $args, $request );
+
+		if ( ! empty( $request['format'] ) ) {
```
note to reviewers: this block appears to be the only part of this method which is changed in this patch.

dmsnell commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,699 @@
+		if ( ! empty( $request['format'] ) ) {
+			// If format is not an array, convert it to an array so that the
+			// required prefix can be added to all items.
+			$formats = is_array( $request['format'] ) ? $request['format'] : array( $request['format'] );
```
could we think of a positive match to convert into an array here? is `$request['format'] === 'database'` a valid input? how could someone look at this line + comment and know what the data type is of a non-array input? how could it be analyzed for correctness or soundness?

```php
$formats = $request['format'];
if ( is_string( $formats ) ) {
	$formats = array( $formats );
}
if ( ! is_array( $formats ) ) {
	// not a valid input, abort processing formats.
}
```

carolinan replied:
I don't understand what you mean.

carolinan replied:
Maybe  @Mamaduka can help translate.

Mamaduka replied:
IIRC, `format` here will always be an array; the schema ensures it. When an invalid parameter is provided, REST API will return `400 Bad Request - Invalid parameter(s): format`.

This conversion here is unnecessary.

TimothyBJacobs replied:
Yep exactly. And if a user passes just a single string, the REST API schema sanitization will take care of making it an array.

dmsnell commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,699 @@
+					'operator' => 'NOT EXISTS',
+				);
+				// Remove the standard format:
+				$formats = array_diff( $formats, array( 'standard' ) );
```
same notes here as in above in `blocks.php`

dmsnell commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,699 @@
+			);
+		}
+
+		if ( post_type_supports( $this->post_type, 'post-formats' ) ) {
```
note to reviewers: this appears to be the only change to this method, apart from a new `@since` tag in the PHPDoc header.

dmsnell commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,699 @@
+
+		if ( post_type_supports( $this->post_type, 'post-formats' ) ) {
+			$query_params['format'] = array(
+				'description' => __( 'Limit result set to items assigned one or more formats.' ),
```
is this limiting to one or more formats, or one or more of the _given_ formats?

carolinan replied:
I don't feel strongly about this description, I only followed the same format that the other descriptions use.

dmsnell replied:
in that case I might suggest diverging from the others since this reads ambiguously. a single word could clarify.

dmsnell commented on lib/compat/wordpress-6.7/post-formats.php:
```diff
@@ -0,0 +1,23 @@
+<?php
+/**
+ * Temporary compatibility shims for changes to post formats in Gutenberg.
```
"Temporary" generally means "permanent" in software unless there are clear conditions listed under which someone can change it. What makes this temporary? What needs to happen to remove it? How should someone remove it when that occurs?

Is this more or less saying this?

```php
/**
 * Shims for changes to post formats until they are merged in all
 * versions of WordPress which Gutenberg supports.
 *
 * @package gutenberg
 */
```

carolinan replied:
As I understand it, it is temporary in the plugin because the compatibility files are removed as soon as possible:

- The change is added to core.
- The plugin has a minimum required WordPress version where the change is implemented.

carolinan replied:
I think this change is not relevant unless you intended to update all the file headers in lib/compat/version.

dmsnell replied:
"Temporary" can be very confusing though. I don't think we have to change existing code to address it. What you wrote in the first reply to my comment would be considerably more helpful to someone who comes around to this code later and wonders why it's still there.

dmsnell commented on lib/compat/wordpress-6.7/post-formats.php:
```diff
@@ -0,0 +1,23 @@
+function gutenberg_post_format_rest_posts_controller( $args ) {
+	/**
+	 * This hook runs before support values are available via `post_type_supports`.
+	 * * Check registration arguments for REST API controller override.
```
nitpick: looks like an extraneous ` *` appeared

carolinan replied:
```suggestion
	 * Check registration arguments for REST API controller override.
```

dmsnell reviewed (commented):
@carolinan thanks for proposing this! I shared a number of thoughts on the PHP code, but everything is up for discussion, so please don't feel like everything I mention is a demand.

carolinan committed:
Merge branch 'trunk' into try/post-format-query

carolinan committed:
gutenberg_filter_query_loop_block_query_vars_post_format: Switch to using an early return with empty().

carolinan committed:
Update lib/compat/wordpress-6.7/post-formats.php

carolinan committed:
End the comment with a full stop.

carolinan commented:
> Now custom post types with formats are not filtered in the editor

This has been resolved 🎉 , thank you @Mamaduka

carolinan committed:
Merge branch 'trunk' into try/post-format-query

carolinan committed:
Update doc blocks and replace aray_diff with unset and array_search.

carolinan committed:
Update doc block

carolinan committed:
Add backport changelog

carolinan committed:
Oops: Fix the copy-paste typo in the changelog entry.

dmsnell commented on lib/compat/wordpress-6.7/blocks.php:
```diff
@@ -43,3 +43,63 @@ function gutenberg_filter_block_type_metadata_settings_allow_variations_php_file
+/**
+ * Adds post format query vars to the query loop block's WP_Query when the block's attributes call for them.
+ *
+ * @see query_loop_block_query_vars
```
I _think_ these need quotes around them to link to the filter, vs. to a function of the same name.

```php
/*
 * @see 'query_loop_block_query_vars'
 */ 
```

carolinan replied:
I think you are right, the PHP documentation standard says
```
Inline @see tags can also be used to auto-link hooks in core:
Hooks, e.g. 'save_post'
Dynamic hooks, e.g. '$old_status_to_$new_status'
```
https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#formatting-guidelines

dmsnell commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,700 @@
+					'operator' => 'NOT EXISTS',
+				);
+				// Remove the standard format, since it cannot be queried.
+				unset( $formats[ array_search( 'standard', $formats, true ) ] );
```
I'm afraid I confused things a bit when I first misunderstood the data type of `$formats`. Given your response, I meant to indicate that your first code was fine with `array_diff()`.

Now, this is also good because we know `standard` is in the array, but otherwise in line 177 we would risk accessing `$formats[ false ]`, which is apparently benign but confusing still.

It's fine to leave this as-is, or revert to your original proposal.

A third option would be to make this entire block aware of the fact that `standard` is at a specific place, but I don't think this is necessary or blocking.

```php
$standard_at = array_search( 'standard', $formats, true );
if( false !== $standard_at ) {
	$tax_query[] = array( … );
	unset( $formats[ $standard_at ] );
}
```

though in review I realize there's another possibility not taken into account, which is that `standard` could appear many times. this seems like an unlikely situation, but if having `standard` in there is a problem, it might be worth considering removing all standard formats in the list. this could be done here, or in a follow-up patch.

at this point, an array filter or `foreach` might be the easier approach for PHP

```php
$has_seen_standard = false;
foreach ( $formats as $index => $format ) {
	if ( 'standard' !== $format ) {
		continue;
	}

	if ( ! $has_seen_standard ) {
		$has_seen_standard = true;
		$tax_query[] = array( … );
	}

	unset( $formats[ $index ] );
}
```

TimothyBJacobs replied:
Handling `standard` multiple times shouldn't be necessary. We can specify the `uniqueItems` keyword on the `format` schema definition.

dmsnell reviewed (commented):
thanks for continuing on this @carolinan!
I left a couple of new notes, but nothing is a major issue.

TimothyBJacobs commented on lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php:
```diff
@@ -0,0 +1,700 @@
+		if ( post_type_supports( $this->post_type, 'post-formats' ) ) {
+			$query_params['format'] = array(
+				'description' => __( 'Limit result set to items assigned one or more given formats.' ),
+				'type'        => 'array',
+				'items'       => array(
```
```suggestion
				'type'        => 'array',
				'uniqueItems' => true,
				'items'       => array(
```

This let's us ensure a user doesn't pass a format multiple times.

carolinan replied:
Thank you!

carolinan committed:
Update lib/compat/wordpress-6.7/class-gutenberg-rest-posts-controller-6-7.php

Co-authored-by: Timothy Jacobs <timothy@ironbounddesigns.com>

carolinan committed:
Rename gutenberg_filter_query_loop_block_query_vars_post_format to gutenberg_add_format_query_vars_to_query_loop_block

carolinan committed:
Gutenberg_REST_Posts_Controller_6_7 get_items: remove the array conversion

$request['format'] does not need to be converted into an array, any array conversion needed is handled by the schema.
This commit removes the array conversion from get_items.

carolinan commented:
I have removed the array conversion in `get_items` in `Gutenberg_REST_Posts_Controller_6_7` and everything seems to be working as expected. Thank you.

---
I tested manually changing the block attributes by editing the markup of the query loop in the block editor.
This is the expected format: `"format":["gallery","video"]`. If I change it to for example `"format":"aside"` there is a fatal PHP error. This would happen independently of the change to get_items, (in blocks.php) and it might be a good idea to check if `format` really is an array and fail more gracefully. Would it be overkill?

Mamaduka commented:
> it might be a good idea to check if format really is an array and fail more gracefully. Would it be overkill?

Yes. Let's add check-in `blocks.php` and return early if `$formats` isn't an array.

carolinan committed:
Add two checks that help fail gracefully when errors occur

In gutenberg_add_format_query_vars_to_query_loop_block: return early if the format attribute is not an array.

In FormatControls: Prevent a JS error that occurs if the format is not an array.
- The purpose is to prevent a user error from breaking the editor, for example if the format is invalid because the user has edited the block markup incorrectly. A check is implemented that preserves the value if for format is an array or string, and uses an empty array if it is neither.

carolinan commented:
I made two changes:
I updated the early return that already existed in blocks.php `gutenberg_add_format_query_vars_to_query_loop_block` to also check if the format is an array.

There was still a JavaScript error in the editor that prevented the editor from functioning, so I also added a check with a
string to array conversion to the `FormatControls`.
- That way when I make a typo in the code editor, (of course this is fiction because that has never happened to me ;) ) both the editor and format control will still work. The result in the query itself will depend on the typo, but it can be corrected by using the control.

carolinan committed:
Update the docBlock for gutenberg_add_format_query_vars_to_query_loop_block

Add apostrophes around the hook name in the docBlock of gutenberg_add_format_query_vars_to_query_loop_block.

Mamaduka commented on packages/block-library/src/query/edit/inspector-controls/format-controls.js:
```diff
@@ -45,6 +45,20 @@ function formatNamesToValues( names, formats ) {
 }
 
 export default function FormatControls( { onChange, query: { format } } ) {
+	// 'format' is expected to be an array. If it is not an array, for example
+	// if a user has manually entered an invalid value in the block markup,
+	// convert it to an array to prevent JavaScript errors.
+	let formatArray = format;
+	if ( Array.isArray( format ) ) {
+		formatArray = format;
+	} else if ( typeof format === 'string' ) {
+		// If the format is a string, preserve it as an array.
+		formatArray = [ format ];
+	} else {
+		// If the format is not a string or an array, set it to an empty array.
+		formatArray = [];
+	}
```
I think this can be oneliner; the editor often does similar normalizations.

```js
const normalizedFormats = Array.isArray( format ) ? format : [ format ];
```

carolinan replied:
What if format is neither a string or array?

carolinan replied:
Well either way it will validated against the available formats later.

Mamaduka replied:
> Well either way it will validated against the available formats later.

Exactly. Also, if someone really wants to break to block, it's probably better to show them an error. Makes debugging easier

carolinan committed:
FormatControls: simplify array conversion

carolinan committed:
Merge branch 'trunk' into try/post-format-query

Mamaduka reviewed (approved):
This is testing well for me, and I couldn't spot any regression to Query block behavior.

carolinan committed:
Merge branch 'trunk' into try/post-format-query

dmsnell commented:
Nicely done, @carolinan!

annezazu commented:
Agreed! Great work 👏🏼

jasmussen commented:
Nice! Made it!

bystrzan commented:
Millions of people asked for it, but you mention only 2 (@justintadlock @SergeyBiryukov) - a closed circle of self-adoration.

carolinan commented:
@bystrzan that is incredibly uncalled for. These contributors are mentioned specifically because they helped out but not by submitting code or reviews **on GitHub.**

All contributors who participate on GitHub are listed **automatically** by the tools that are used in this repository. Other contributors need to be added manually. **Which is why I literally wrote Additional props.**

This type of comment is uncalled for, rude, and an attempt to bring negativity and distract people from contributing to WordPress.
If what you were trying to say instead was that "this (name) is missing props for this commit", then please do that.
